### DISCUSSION TITLE
Terraform cloud private modules

### DISCUSSION THREAD

**Discussion by KevinM2k** [renovatebot/renovate#21258] [Request Help] [answered]
i'm currently trying to pull out some private terraform modules and i'm getting 401 errors, it seems the only way is by using bearer tokens. I've tried this:
```
"hostRules": [ { 
   "matchHost": "https://app.terraform.io", 
   "token": "${ENV_TFE_TOKEN}", 
   "authType": "Bearer"
 } ] 
```
But this doesn't work. The URL it is hitting returns a 401, and it does on my local machine unless I pass that same token as a bearer token to the request.

This is the error:
```
DEBUG: GET https://app.terraform.io/api/registry/v1/modules/<my-private-org>/<my-module>/versions = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=29) (repository=my-repo) DEBUG: Datasource unauthorized (repository=my-repo) "datasource": "terraform-module", "packageName": "app.terraform.io/<my-private-org>/<my-module>", "url": "https://app.terraform.io/api/registry/v1/modules/<my-private-org>/<my-module>/versions" 
```
If I do:
```
curl -H"Authorization: Bearer <token>" https://app.terraform.io/api/registry/v1/modules/<my-private-org>/<my-module>/versions 
```
Its successful, without the Bearer I get a 401.

**Comment by viceice** [comment, marked as answer]
hostrules seems wrong, renovate doesn't support any variable replacement inside renovate config files

**Reply by KevinM2k** [reply]
How do you authenticate with private registries?  From what I can see on the docs this is the recommended way unless I'm missing something.

**Reply by viceice** [reply]
- Are you using the hosted github app or do you self-host renovate?
- How do you run renovate (docker/ helm / npm) if you selfhost?

**Reply by KevinM2k** [reply]
Use gitlab, and it's self hosted in a kubernetes cluster, runs as a cron job and the renovate config is passed as an env var in the ConfigMap.

**Reply by viceice** [reply]
Then something is wrong when passing the token `authType` can be removed, as `Bearer` is the default value.

You can add a `config.js` with this content to test the value

```js
console.log(process.env.ENV_TFE_TOKEN?.substring(0,2));
module.exports = {};
```

so it will print the first two chars of the token to validate it's passed correctly

**Reply by KevinM2k** [reply]
Does the config.js file need to go in to the individual respositories? I do think you might be on to something that the token isn't passing correctly. The env var is in gitlab ci, I'm wondering whether I've done the env vars correctly.

I know bearer is default btw, I passed it in just in case.

**Reply by viceice** [reply]
no, into the renovate pod

**Reply by viceice** [reply]
renovate doesn't read repo environment variables. 

- so you need to add it plain text to repo config 
- or add it to global bot config as hostrule or secret (can be repo specific via `repositories` option)

**Reply by KevinM2k** [reply]
Ok so I create that config.js, and I put the hostrules in that, i use process.env in the config file to get the env var and I mount that file in to the pod. 

Does host rules go directly in to the modules.export object?

I take it that config.js then overwrites the defaults and allows you to set things up that the Devs don't need to worry about within there own config.

**Reply by rarkins** [reply]
Yes

**Reply by KevinM2k** [reply]
Just to clarify, this worked:
```
config.js
----------
module.exports = {
   "hostRules": [
     { 
        "matchHost": "app.terraform.io",
        "token": process.env.TFE_TOKEN,
      },
    ],
};
``` 

thanks for the help @viceice

**Comment by Mirdrack** [comment]
What should be the proper way to do this with the Github App?
How can I keep my secrets encrypted?

**Reply by viceice** [reply]
https://docs.renovatebot.com/getting-started/private-packages/#mend-renovate-hosted-app-encryption

this is a better link
